python/pyarrow/_fs.pyx

```diff
@@ -269,6 +269,15 @@ cdef class FileSystem:
     cdef inline shared_ptr[CFileSystem] unwrap(self) nogil:
         return self.wrapped
 
+    def equals(self, FileSystem other):
+        return self.fs.Equals(other.unwrap())
+
+    def __eq__(self, other):
+        try:
+            return self.equals(other)
+        except TypeError:
+            return NotImplemented
```

Comment:
I think this check is unnecessary; C++ should fail to compile if a subclass of CFileSystem did not implement Equals()

Reply:
This is python and cython specific. 

`self.equals` raises a `TypeError` if the passed argument is not an instance of FileSystem, but the object needs to be comparable with other objects without raising an error:

```py
fs == 'any-non-filesystem-objects'  # should return false instead of raising TypeError
```

`NotImplemented` is a [special exception](https://docs.python.org/3.6/library/constants.html#NotImplemented) to let the other operand to override the comparison operator.